### PR TITLE
State timers documentation

### PR DESCRIPTION
**Pull request by reuvenlax**

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @dpmills 
R: @rezarokni

File: website/src/documentation/programming-guide.md [hunk 1]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+    state.add(1);
+  }
+}));
+```
+
+#### BagState
+A common use case for state is to accumulate multiple elements. `BagState` allows for accumulating elements in an 
+unordered manner. This allows for addition of elements to the collection without requiring the reading of the entire
+collection first, which is an efficiency gain. In addition, runners that support paged reads can allow individual
+bag collections larger than available memory.
```
Comment (dpmills) [inline website/src/documentation/programming-guide.md]: bag collections seems redundant; maybe just "bags"?

File: website/src/documentation/programming-guide.md [hunk 2]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+      // Occasionally we fetch and process the values.
+      Iterable<ValueT> values = state.read();
+      processValues(values);
+      state.clear();  // Clear the state for this key.
+    }
+  }
+}));
+```
+### 10.2 Deferred state reads {#deferred-state-reads}
+When a `DoFn` contains multiple state specifications, reading each one in order can be slow. Calling the read() function
```
Comment (dpmills) [inline website/src/documentation/programming-guide.md]: put code in backticks

File: website/src/documentation/programming-guide.md [hunk 3]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+
+```java
+PCollection<KV<String, ValueT>> perUser = readPerUser();
+perUser.apply(ParDo.of(new DoFn<KV<String, ValueT>, OutputT>() {
+  @TimerId("timer") private final TimerSpec timer = TimerSpecs.timer(TimeDomain.PROCESSING_TIME);
+
+  @ProcessElement public void process(@TimerId("timer") Timer timer) {
+     ...
+     // Set a timer to go off 30 seconds in the future.
+     timer.offset(Duration.standardSeconds(30).setRelative();
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: Unfortunately today the timer is set in the setRelative() method, so the subsequent offset() does not do anything!

[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: Also setRelative() is a void method :)

File: website/src/documentation/programming-guide.md [hunk 4]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
   }
 }
 ```  
+
+## 10. State and Timers {#state-and-timers}
+Beam's windowing and triggering facilities provide a powerful abstraction for grouping and aggregating unbounded input
+data based on timestamps. However there are aggregation use cases for which Beam's windows and triggers are not the best
```
Comment (rezarokni) [inline website/src/documentation/programming-guide.md]: However there are aggregation use cases for which developers may require a higher degree of control...

File: website/src/documentation/programming-guide.md [hunk 5]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+fit. Beam provides an API for manually managing per-key state, allowing for finer-grained control over aggregations.
+
+Beam's state API models state per key. To use the state API, you start out with a keyed `PCollection`, which in Java
+is modeled as a `PCollection<KV<K, V>>`. A `ParDo` processing this `PCollection` can now declare state variables. Inside
+the `ParDo` these state variables can be used to write or update state for the current key or to read previous state
+written for that key. State is always fully scoped only to the current processing key.
+
+Windowing can still be used together with stateful processing. All state for a key is scoped to the current window. This
```
Comment (rezarokni) [inline website/src/documentation/programming-guide.md]: Also important to point out that you can use different window strategies before the State call. Something which I have seen folks not realise until pointed out.
For example Window+Combiner as a reducer before data comes to the state DoFn

File: website/src/documentation/programming-guide.md [hunk 6]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+### 10.1 Types of state {#types-of-state}
+Beam provides several types of state:
+
+#### ValueState
+A DoFn declares states to be accessed by creating final `StateSpec` member variables representing each state. Each
+state must be named using the `StateId` annotation; this name is unique to a ParDo in the graph and has no relation
+to other nodes in the graph. A `DoFn` can declare multiple state variables.
+
+If the type of the ValueState has a coder registered, then Beam will automatically infer the coder for the state value.
+For example, the following ParDo creates a single state variable that accumulates the number of elements seen.
```
Comment (rezarokni) [inline website/src/documentation/programming-guide.md]: Worth explicitly calling out the lack of order? Although it doesnt matter for this use case, ValueState can be misused for other things when not aware of no order

File: website/src/documentation/programming-guide.md [hunk 7]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+  }
+  
+   @OnTimer("timer") public void onTimer() {
+      //Process timer.
+   }
+}));
+
+```
+#### 10.3.2 Processing-time timers {#processing-time-timers}
+Processing-time timers fire when the real wall-clock time passes. This is often used to create larger batches of data
```
Comment (rezarokni) [inline website/src/documentation/programming-guide.md]: Worth mentioning expected behaviour if used in batch pipeline. If I recall direct runner doesnt fire process time timers?

File: website/src/documentation/programming-guide.md (hunk 6 above)
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md (hunk 5 above)
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md (hunk 4 above)
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: I like it. done

File: website/src/documentation/programming-guide.md (hunk 2 above)
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md (hunk 1 above)
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 8]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+   @OnTimer("timer") public void onTimer() {
+      //Process timer.
+   }
+}));
+
+```
+
+#### 10.3.3 Dynamic timer tags {#dynamic-timer-tags}
+Beam also supports dynamically setting a timer tag using `TimerMap`. This allows for setting multiple different timers
+in a `ParDo` and allowing for the timer tags to be dynamically chosen - e.g. based on data in the input elements. A
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 9]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+PCollection<KV<String, ValueT>> perUser = readPerUser();
+perUser.apply(ParDo.of(new DoFn<KV<String, ValueT>, OutputT>() {
+  @StateId("elementBag") private final StateSpec<BagState<ValueT>> elementBag = StateSpecs.bag();
+  @StateId("timerSet") private final StateSpec<ValueState<Boolean>> timerSet = StateSpecs.value();
+  @TimerId("outputState") private final TimerSpec timer = TimerSpecs.timer(TimeDomain.PROCESSING_TIME);
+  
+  @ProcessElement public void process(
+      @Element KV<String, ValueT> element, 
+      @StateId("elementBag") BagState<ValueT> elementBag,
+      @StateId("timerSet") ValueState<Boolean> timerSet,
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 10]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+performance. There are two common strategies for garbage collecting state.
+
+##### 10.4.1 Using windows for garbage collection {#using-windows-for-garbage-collection}
+All state and timers for a key is scoped to the window it is in. This means that depending on the timestamp of the 
+input element the ParDo will see different values for the state depending on the window that element falls into. In
+addition, once the input watermark passes the end of the window, the runner should garbage collect all state for that
+window. (note: if allowed lateness is set to a positive value for the window, the runner must wait past the end of the
+window in order to garbage collect state). This can be used as a garbage-collection strategy.
+
+For example, give the following:
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 11]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+### 10.4 Garbage collecting state {#garbage-collecting-state}
+Per-key state needs to be garbage collected, or eventually the increasing size of state may negatively impact 
+performance. There are two common strategies for garbage collecting state.
+
+##### 10.4.1 Using windows for garbage collection {#using-windows-for-garbage-collection}
+All state and timers for a key is scoped to the window it is in. This means that depending on the timestamp of the 
+input element the ParDo will see different values for the state depending on the window that element falls into. In
+addition, once the input watermark passes the end of the window, the runner should garbage collect all state for that
+window. (note: if allowed lateness is set to a positive value for the window, the runner must wait past the end of the
+window in order to garbage collect state). This can be used as a garbage-collection strategy.
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 12]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+#### 10.3.4 Timer output timestamps {#timer-output-timestamps}
+By default, event-time timers will hold the output watermark of the `ParDo` to the timestamp of the timer. This means
+that if a timer is set to 12pm, any windowed aggregations or event-time timers later in the pipeline graph that finish  
+after 12pm will not expire. The timestamp of the timer is also the default output timestamp for the timer callback. This
+means that any elements output from the onTimer method will have a timestamp equal to the timestamp of the timer firing.
+For processing-time timers, the default output timestamp and watermark hold is the value of the input watermark at the
+time the timer was set.
+
+In some cases, a pipeline needs to output timestamps earlier than the timer expiration time, and therefore also needs to
+hold the watermark to those timestamps. For example, consider the following pipeline that temporarily batches records
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 13]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+
+#### 10.3.4 Timer output timestamps {#timer-output-timestamps}
+By default, event-time timers will hold the output watermark of the `ParDo` to the timestamp of the timer. This means
+that if a timer is set to 12pm, any windowed aggregations or event-time timers later in the pipeline graph that finish  
+after 12pm will not expire. The timestamp of the timer is also the default output timestamp for the timer callback. This
+means that any elements output from the onTimer method will have a timestamp equal to the timestamp of the timer firing.
+For processing-time timers, the default output timestamp and watermark hold is the value of the input watermark at the
+time the timer was set.
+
+In some cases, a pipeline needs to output timestamps earlier than the timer expiration time, and therefore also needs to
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: fixed paren

File: website/src/documentation/programming-guide.md [hunk 14]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+    // The runner can now batch all three states into a single read, reducing latency.
+    processState1(state1.read());
+    processState2(state2.read());
+    processState3(state3.read());
+  }
+}));
+```
+
+### 10.3 Timers {#timers}
+Beam provides a per-key timer callback API. This allows for delayed processing of data aggregated using the state API.
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 15]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+
+
+### 10.1 Types of state {#types-of-state}
+Beam provides several types of state:
+
+#### ValueState
+A DoFn declares states to be accessed by creating final `StateSpec` member variables representing each state. Each
+state must be named using the `StateId` annotation; this name is unique to a ParDo in the graph and has no relation
+to other nodes in the graph. A `DoFn` can declare multiple state variables.
+
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

File: website/src/documentation/programming-guide.md [hunk 16]
```diff
@@ -3121,3 +3121,515 @@ public class MyMetricsDoFn extends DoFn<Integer, Integer> {
...
+Windowing can still be used together with stateful processing. All state for a key is scoped to the current window. This
+means that the first time a key is seen for a given window any state reads will return empty, and that a runner can
+garbage collect state when a window is completed.  Merging windows are not currently supported when using state and timers.
+
+
+### 10.1 Types of state {#types-of-state}
+Beam provides several types of state:
+
+#### ValueState
+A DoFn declares states to be accessed by creating final `StateSpec` member variables representing each state. Each
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply website/src/documentation/programming-guide.md]: done

**Pull request merged**